File: src/mlogix/compiler/type/Type.java
package mlogix.compiler.type;

import arc.struct.ObjectMap;

public class Type {
    public final String name;
    public final ObjectMap<String, Type> fields = new ObjectMap<>();
    public final ObjectMap<String, Type> methods = new ObjectMap<>();

    public Type(String name) {
        this.name = name;
    }

    public Type addField(String name, Type type){
        fields.put(name, type);
        return this;
    }

    public Type addMethod(String name, Type fn){
        methods.put(name, fn);
        return this;
    }
}

File: src/mlogix/compiler/type/Type.kt
package mlogix.compiler.type

import arc.struct.ObjectMap

open class Type(val name: String) {
    val fields: ObjectMap<String, Type> = ObjectMap<String, Type>()
    val methods: ObjectMap<String, FunctionType> = ObjectMap<String, FunctionType>()

    fun addField(name: String?, type: Type?): Type {
        fields.put(name, type)
        return this
    }

    fun addMethod(name: String?, fn: FunctionType?): Type {
        methods.put(name, fn)
        return this
    }
}